### ISSUE TITLE
Using online connection and tiles that aren't available

### ISSUE THREAD

**Issue by spyhunter99** [osmdroid/osmdroid#131]
I've noticed that if osmdroid is pointed at an online map server and you're zoomed or panned to an area in which the map tiles do not exist, osmdroid continues to request the same tile and over again. Is this expected behavior? Eventually, it seems like fill up a queue or something then causes the overall performance of osmdroid to drop like a rock. Maybe we can add an expiring blacklist of tiles or something to that affect to prevent extraneous http requests?

**Comment by neilboyd** [comment]
Probably the best fix is for `MapTileDownloader` to return a 404 drawable when `line.getStatusCode() == 404`

**Comment by spyhunter99** [comment]
i added an attempted download count to the MapTile class, however this approach doesn't work since the tile looper requests tiles for everything that on screen that isn't visible.

**Comment by monsieurtanuki** [comment]
I guess it's fixed now, isn't it?

**Comment by spyhunter99** [comment]
@monsieurtanuki i think that the tile approximator probably fixes this. My issue was that the tile downloader would continuously try to download tiles when the device was connected to a network, but not the internet. The "am i online" checks were true, but there was no route the map server. The constant looping and connection reattempts kill the battery quickly.

I think the fix is a LRU cache of sorts that keeps track of the most recent set of tiles that have been attempted for download recently and increment a counter every time it fails. past a given threshold, give up

**Comment by monsieurtanuki** [comment]
> connected to a network, but not the internet

What we need in osmdroid is access to the internet, and never only to the network, right? Then I think we shouldn't care about the network being on/off, just the internet on/off. And the "am i online" and `isDowngradedMode` give the correct "I cannot download anything as I have no connection to external data" answer.

**Comment by spyhunter99** [comment]
the test needs to be can i reach map server X, which may be on the internet but may not be.

**Comment by spyhunter99** [comment]
source history says it was added july 22 this year, last release was july 1st. javadoc is wrong, sorry!

**Comment by monsieurtanuki** [comment]
@spyhunter99 I must say I'm a bit lost: could you be more explicit about trying to reach a map server that is NOT on the internet? Something like a Linux server file system I can reach from my smartphone?

**Comment by spyhunter99** [comment]
There are use cases whereby you have an isolated network that is not internet facing. On this isolated network, is a map tile server that your osmdroid powered app can reach. However, this tile server only has tiles to zoom 15 and the user (you) are zooming to zoom 18. At the time this was reported, osmdroid was continuously requesting all zoom 18 tiles in a never ending loop. As you can imagine, this was resource intensive, causing high cpu usage and battery drain and in general, poor performance. I'm not sure if this is still an issue or not but i suspect that it is

**Comment by pasniak** [comment]
When I zoom in to 15 (on a XYTileSource with maxZoom 14), I observe high CPU usage (on an emulator) due to getNetworkAvailable, more precisely
```
final NetworkInfo networkInfo = mConnectionManager.getActiveNetworkInfo();
```
![screenshot from 2018-02-23 05-35-55](https://user-images.githubusercontent.com/19778666/36592156-9bb27654-1862-11e8-965f-6ae993a18b8a.png)

Could that getActiveNetworkInfo call should be throttled somehow? (I actually need to zoom beyond 14 as I want to make visible a more detailed GroundOverlay2)

On my tablet phone hardware this high CPU usage does not happen but is quite annoying on my PC emu (AS) as my PC starts a noisy CPU fan - a great feature to detect CPU drains :-)

84938796ed097a216c5eba01e56b14e54c19822c as of 2/23/2018

**Comment by pasniak** [comment]
correction: this also happens on my tablet/phone (I was looking at the wrong profiler window...)

**Comment by monsieurtanuki** [comment]
Allegedly fixed by #1107. Feel free to reopen.
